### ISSUE TITLE
Improve recursive type checking

### ISSUE THREAD

**Issue by danielnixon** [danielnixon/eslint-plugin-total-functions#732]
The "five turtles" solution from https://github.com/danielnixon/eslint-plugin-total-functions/issues/192 abruptly terminates all recursive type checking.

Prior to that change, linting this would never terminate:

```ts
type Foo<U> = {
  readonly a: U;
  readonly b: Bar<Bar<U>>;
};

type Bar<U> = {
  readonly a: U;
  readonly b: Foo<Foo<U>>;
};

declare const foo: Foo<string>;

const bar: Bar<string> = foo;
```

It turns out that solution is overzealous or overly-naive, because it now means that checking nested structures terminates abruptly in a way that doesn't catch this assignment from mutable to readonly:

```ts


type A = {
  readonly a: string;
};

type B = {
  a: A;
};

type C = {
  b: B;
};

type D = {
  c: C;
};

type E = {
  d: D;
};

type F = {
  e: E;
};

type G = {
  f: F;
};

type A2 = {
  a: string;
};

type B2 = {
  a: A2;
};

type C2 = {
  b: B2;
};

type D2 = {
  c: C2;
};

type E2 = {
  d: D2;
};

type F2 = {
  e: E2;
};

type G2 = {
  f: F2;
};

declare const g: G;

// Not flagged, should be
export const g2: G2 = g;
```

If we change that first type to

```ts
type A = {
  a: boolean;
};
```

we see TypeScript itself is still able to see the issue in the deeply nested structure:

```
Type 'G' is not assignable to type 'G2'.
  The types of 'f.e.d.c.b.a.a' are incompatible between these types.
    Type 'boolean' is not assignable to type 'string'.
```

So there must be some improvement we can make that catches this and doesn't fall prey to infinite recursion. What is it? We might have to go spelunking in TypeScript's source.

We know the issue only arises when generics are involved so to start with, we can probably decrement our counter only when we're traversing from one generic type to another. And further, we can probably reset the counter when traversing from a non-generic type to a generic one.

Although, we'd have to be very careful about deciding whether a type is generic or not.

**Comment by danielnixon** [comment]
Related: https://github.com/RebeccaStevens/is-immutable-type/issues/98

**Comment by danielnixon** [comment]
`getRecursionIdentity` and `isDeeplyNestedType` from `checker.ts` look like they hold the answer.
